packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, named, optional } from 'inversify';
+import { ILogger } from '../common/logger';
+import { AbstractDialog } from './dialogs';
+import { MessageService } from '../common/message-service';
+import { ContributionProvider } from '../common/contribution-provider';
+import { StatusBar, StatusBarAlignment } from './status-bar/status-bar';
+import { FrontendApplicationContribution } from './frontend-application';
+import { ConnectionState, ConnectionStatusChangeEvent, ConnectionStatusContribution } from '../common/connection-status';
+
+@injectable()
+export class ConnectionStatusOptions {
+
+    static DEFAULT: ConnectionStatusOptions = {
+        requestTimeout: 1000,
+        retry: 5,
+        retryInterval: 2000,
+    };
+
+    /**
+     * Timeout for the HTTP GET request in milliseconds.
+     */
+    readonly requestTimeout: number;
+
+    /**
+     * Number of accepted timeouts.
+     */
+    readonly retry: number;
+
+    /**
+     * Retry interval in milliseconds.
+     */
+    readonly retryInterval: number;
+
+}
+
+@injectable()
+export class FrontendConnectionStatusService implements FrontendApplicationContribution {
+
+    private connectionState: ConnectionStateMachine;
+    // tslint:disable-next-line:no-any
+    private timer: any | undefined;
+
+    constructor(
+        @inject(ConnectionStatusOptions) @optional() protected readonly options: ConnectionStatusOptions = ConnectionStatusOptions.DEFAULT,
+        @inject(ContributionProvider) @named(ConnectionStatusContribution) protected readonly contributionProvider: ContributionProvider<ConnectionStatusContribution>
+    ) {
+        this.connectionState = new ConnectionStateMachine({ threshold: this.options.retry });
+    }
+
+    onStart() {
+        this.timer = setInterval(() => {
+            const xhr = new XMLHttpRequest();
+            xhr.timeout = this.options.requestTimeout;
+            xhr.onreadystatechange = event => {
+                const { readyState, status } = xhr;
+                if (readyState === XMLHttpRequest.DONE) {
+                    const success = status === 200;
+                    this.connectionState = this.connectionState.next(success);
+                    this.contributionProvider.getContributions().forEach(contribution => contribution.onStatusChange(this.connectionState));
+                }
+            };
+            xhr.onerror = () => {
```

Comment:
i see that `xhr` has other events, like `ontimeout`, should not we handle them as well?

Reply:
OK.